### PR TITLE
Disable fast math

### PR DESCRIPTION
**Pull request by csarofeen**
Disable fast math to be more consistent with eager mode while trying to keep as much perf as possible. Some regressions exist, but uncertain if there's any way around them.

image

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/runtime/helpers.cu [hunk 1]
```diff
@@ -279,3 +279,19 @@ template <>
 double pow<double>(double a, double b) {
   return ::pow(a, b);
 }
+
+float pow(float a, int b) {
+  return pow(a, (float)b);
+}
+
+float pow(double a, int b) {
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/runtime/helpers.cu]: Shouldn't the return type be `double`?

File: torch/csrc/jit/codegen/cuda/runtime/helpers.cu [hunk 2]
```diff
@@ -279,3 +279,19 @@ template <>
...
+  return pow(a, (double)b);
+}
+
+float pow(float a, int64_t b) {
+  return pow(a, (float)b);
+}
+
+float pow(double a, int64_t b) {
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/runtime/helpers.cu]: Same here.

File: torch/csrc/jit/codegen/cuda/ops/normalization.cpp [hunk 3]
```diff
@@ -23,7 +23,7 @@ TensorView* softmax(TensorView* x, int dim) {
   auto exp_val = exp(x_max_sub);
   auto sum_exp = sum(exp_val, {kReductionAxis});
   auto bcast_sum = broadcast(sum_exp, broadcast_mask);
-  auto y = div(exp_val, bcast_sum);
+  auto y = mul(exp_val, reciprocal(bcast_sum));
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/ops/normalization.cpp:26]: Would it make sense to have this transformation somewhere during the lowering?

**Review by naoyam — COMMENTED** [review]
Some minor questions

File: torch/csrc/jit/codegen/cuda/runtime/helpers.cu (hunk 1 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/runtime/helpers.cu]: Should we just have
```
template<typename T0,
         typename T1,
         typename = std::enable_if_t<std::is_floating_point<T0>::value>,
         typename = std::enable_if_t<!std::is_same<T0, T1>::value>>
T0 pow(T0 a, T1 b) {
  return pow(a, (T0) b);
}
```

**Review by jjsjann123 — APPROVED** [review]
LGTM

File: torch/csrc/jit/codegen/cuda/ops/normalization.cpp (hunk 3 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/ops/normalization.cpp:26]: I don't know if we enable this everywhere. I was trying to be a bit more conservative. We could just overwrite arith.cpp::div if we wanted.

File: torch/csrc/jit/codegen/cuda/runtime/helpers.cu (hunk 1 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/runtime/helpers.cu]: Can we import type traits in nvrtc? Leaving as is for now but feel free to follow up on it.

**Pull request merged**